code/rewriter.go

```diff
@@ -182,22 +121,60 @@ func (r *Rewriter) rewriteIfStmt(v *ast.IfStmt) error {
 	return nil
 }
 
-func (r *Rewriter) rewriteExprs(exprs []ast.Expr) error {
-	for _, expr := range exprs {
+func (r *Rewriter) rewriteExpr(expr ast.Expr) error {
+	switch ex := expr.(type) {
+	case *ast.FuncLit:
 		// return func(){...},
-		if fn, ok := expr.(*ast.FuncLit); ok {
+		return r.rewriteFuncLit(ex)
+
+	case *ast.CallExpr:
+		// return func() int {...}()
+		if fn, ok := ex.Fun.(*ast.FuncLit); ok {
 			err := r.rewriteFuncLit(fn)
 			if err != nil {
 				return err
 			}
 		}
-		// return func() int {...}()
-		if fn, ok := expr.(*ast.CallExpr); ok {
-			err := r.rewriteCallExpr(fn)
-			if err != nil {
-				return err
+
+		for _, arg := range ex.Args {
+			if fn, ok := arg.(*ast.FuncLit); ok {
+				err := r.rewriteFuncLit(fn)
+				if err != nil {
+					return err
+				}
 			}
 		}
+
+	case *ast.UnaryExpr:
+		return r.rewriteExpr(ex.X)
+
+	case *ast.BinaryExpr:
+		// a && func() bool {...} ()
+		// func() bool {...} () && a
+		// func() bool {...} () && func() bool {...} () && a
+		// func() bool {...} () && a && func() bool {...} () && a
+		err := r.rewriteExpr(ex.X)
+		if err != nil {
+			return err
+		}
+		return r.rewriteExpr(ex.Y)
+
+	case *ast.ParenExpr:
+		return r.rewriteExpr(ex.X)
+
+	case *ast.TypeAssertExpr:
+		return r.rewriteExpr(ex.X)
+	}
```

Comment:
Add a `default:` case for the unhandled types? 

I'm also not sure if you need to care about `IndexExpr` (`x[y]`), `StarExpr` (`*x`) etc.

Reply:
addressed